samples/nrf9160/modem_shell/src/sock/sock.c

```diff
@@ -0,0 +1,1320 @@
+		/* Send given amount of data */
+		uint32_t send_buffer_size = SOCK_SEND_BUFFER_SIZE_TCP;
+
+		/* Interval is not supported with data length */
+		if (interval != SOCK_SEND_DATA_INTERVAL_NONE) {
+			shell_error(
+				shell_global,
+				"Data length and interval cannot be specified at the same time");
+			return -EINVAL;
+		}
+
+		if (buffer_size != SOCK_BUFFER_SIZE_NONE) {
+			send_buffer_size = buffer_size;
+		} else if (socket_info->type == SOCK_DGRAM) {
+			send_buffer_size = SOCK_SEND_BUFFER_SIZE_UDP;
+		}
+		if (!sock_send_buffer_calloc(socket_info, send_buffer_size)) {
+			return -ENOMEM;
+		}
+
+		shell_print(
+			shell_global,
+			"Sending %d bytes of data with buffer_size=%d, blocking=%d",
+			random_data_length,
+			send_buffer_size,
+			blocking);
+
+		/* Warn about big buffer sizes as lower levels gets stuck when buffer size
+		 * increases. Not necessarily right above these values you cannot use much bigger
+		 * send buffer.
+		 */
+		if ((socket_info->type == SOCK_STREAM && send_buffer_size > 4096) ||
+		    (socket_info->type == SOCK_DGRAM && send_buffer_size > 1200)) {
+			shell_warn(
+				shell_global,
+				"Sending %d bytes of data with buffer_size=%d, blocking=%d",
+				random_data_length, send_buffer_size, blocking);
+		}
+
+		socket_info->send_bytes_sent = 0;
+		socket_info->send_bytes_left = random_data_length;
+		socket_info->log_receive_data = false;
+		socket_info->send_print_interval = 10;
+		/* Set requested blocking mode for duration of data sending */
+		set_sock_blocking_mode(socket_info->fd, blocking);
+
+		memset(socket_info->send_buffer, 'd', socket_info->send_buffer_size);
+
+		socket_info->start_time_ms = k_uptime_get();
+
+		sock_send_random_data_length(socket_info);
+
+		/* Keep default mode of the socket in non-blocking mode */
+		set_sock_blocking_mode(socket_info->fd, false);
+
+	} else if (interval != SOCK_SEND_DATA_INTERVAL_NONE) {
+
+		if (interval == 0) {
+			/* Stop periodic data sending */
+			if (k_timer_remaining_get(&socket_info->send_info.timer) > 0) {
+
+				k_timer_stop(&socket_info->send_info.timer);
+				shell_print(shell_global, "Socket data send periodic stop");
+			} else {
+				shell_error(shell_global,
+					    "Socket data send stop: periodic data not started");
+				return -EINVAL;
+			}
+		} else if (interval > 0) {
+			/* Send data with given interval */
+
+			/* Data to be sent must also be specified */
+			if (data_out_length < 1) {
+				shell_error(shell_global,
+					    "Data sending interval is specified without "
+					    "data to be sent");
+				return -EINVAL;
+			}
+
+			if (!sock_send_buffer_calloc(socket_info, data_out_length)) {
+				return -ENOMEM;
+			}
+			memcpy(socket_info->send_buffer, data_out, data_out_length);
+			socket_info->send_buffer_size = data_out_length;
+
+			socket_info->send_info.data_format_hex = data_format_hex;
+			shell_print(
+				shell_global,
+				"Socket data send periodic with interval=%d",
+				interval);
+			k_timer_init(
```

Comment:
It's probably good enough for now, but you're simply re-inventing a delayed work here, probably could use it instead in the future.

Reply:
I'll see later in the week if I have time to refactor this

Reply:
I've done this now.